instrumentation/rocketmq/rocketmq-client/rocketmq-client-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/rocketmqclient/v5_0/RocketMqInstrumenterFactory.java

```diff
@@ -41,12 +46,63 @@ public static Instrumenter<PublishingMessageImpl, SendReceiptImpl> createProduce
             .addAttributesExtractor(RocketMqProducerAttributeExtractor.INSTANCE)
             .setSpanStatusExtractor(
                 (spanStatusBuilder, message, sendReceipt, error) -> {
-                  if (null != error) {
+                  if (error != null) {
+                    spanStatusBuilder.setStatus(StatusCode.ERROR);
+                  }
+                });
+    return instrumenterBuilder.buildProducerInstrumenter(MessageMapSetter.INSTANCE);
+  }
+
+  public static Instrumenter<MessageQueueImpl, List<MessageView>> createConsumerReceiveInstrumenter(
+      OpenTelemetry openTelemetry, List<String> capturedHeaders) {
+    RocketMqConsumerReceiveAttributeGetter getter = RocketMqConsumerReceiveAttributeGetter.INSTANCE;
+    MessageOperation operation = MessageOperation.RECEIVE;
+
+    MessagingAttributesExtractor<MessageQueueImpl, List<MessageView>> attributesExtractor =
+        buildMessagingAttributesExtractor(getter, operation, capturedHeaders);
+
+    InstrumenterBuilder<MessageQueueImpl, List<MessageView>> instrumenterBuilder =
+        Instrumenter.<MessageQueueImpl, List<MessageView>>builder(
+                openTelemetry,
+                INSTRUMENTATION_NAME,
+                MessagingSpanNameExtractor.create(getter, operation))
+            .addAttributesExtractor(attributesExtractor)
+            .setSpanStatusExtractor(
+                (spanStatusBuilder, messageView, unused, error) -> {
+                  if (error != null) {
                     spanStatusBuilder.setStatus(StatusCode.ERROR);
                   }
                 });
+    return instrumenterBuilder.buildInstrumenter(SpanKindExtractor.alwaysConsumer());
+  }
+
+  public static Instrumenter<MessageView, ConsumeResult> createConsumerProcessInstrumenter(
+      OpenTelemetry openTelemetry, List<String> capturedHeaders) {
+    RocketMqConsumerProcessAttributeGetter getter = RocketMqConsumerProcessAttributeGetter.INSTANCE;
+    MessageOperation operation = MessageOperation.PROCESS;
+
+    MessagingAttributesExtractor<MessageView, ConsumeResult> attributesExtractor =
+        buildMessagingAttributesExtractor(getter, operation, capturedHeaders);
+
+    SpanLinksExtractor<MessageView> spanLinksExtractor =
```

Comment:
Some backends don't support span links (actually I don't know if anything besides Jaeger supports them), because of that we disable the consumer instrumentation based on a flag (see [rabbitmq](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/6f9ca0b80bb732100fca157e989f9b0da3c93f5a/instrumentation/rabbitmq-2.7/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/rabbitmq/RabbitSingletons.java#L83)).  When this flags is disabled instrumentations (see [kafka](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/6f9ca0b80bb732100fca157e989f9b0da3c93f5a/instrumentation/kafka/kafka-clients/kafka-clients-common/library/src/main/java/io/opentelemetry/instrumentation/kafka/internal/KafkaInstrumenterFactory.java#L79)) produce a single trace where process span is the child of the producer span. Also I think we put the span link on the receive span not the process span, check other messaging instrumentations.

Reply:
OK, I will add the part about the receive instrumentation flag.

As for the span link on the receive span, I am confused with it. All message in RocketMQ is received in batch mode, according to the https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-receiving, the span link should be put on the process span.